polygeist/tools/cgeist/Lib/clang-mlir.cc

```diff
@@ -5127,40 +5121,49 @@ mlir::FunctionOpInterface MLIRASTConsumer::GetOrCreateMLIRFunction(
            FunctionDecl::TemplatedKind::
                TK_DependentFunctionTemplateSpecialization);
     if (ShouldEmit) {
-      functionsToEmit.emplace_back(*Def, Context);
+      LLVM_DEBUG(llvm::dbgs()
+                 << __LINE__ << ": Pushing " << F.getContext() << " function "
+                 << Def->getNameAsString() << " to functionsToEmit\n");
+      functionsToEmit.emplace_back(*Def, F.getContext());
     }
   } else if (ShouldEmit) {
     emitIfFound.insert(name);
   }
-  checkFunctionParent(function, Context, module, deviceModule);
+  checkFunctionParent(function, F.getContext(), module, deviceModule);
   return function;
 }
 
 void MLIRASTConsumer::run() {
   while (functionsToEmit.size()) {
-    auto &F = functionsToEmit.front();
-    const FunctionDecl *FD = &F.decl;
-
+    FunctionToEmit &F = functionsToEmit.front();
     functionsToEmit.pop_front();
 
+    const FunctionDecl *FD = &F.getDecl();
+
     assert(FD->getBody());
     assert(FD->getTemplatedKind() != FunctionDecl::TK_FunctionTemplate);
     assert(FD->getTemplatedKind() !=
            FunctionDecl::TemplatedKind::
                TK_DependentFunctionTemplateSpecialization);
 
-    std::string name;
-    MLIRScanner::getMangledFuncName(name, FD, CGM);
+    std::string mangledName;
+    MLIRScanner::getMangledFuncName(mangledName, FD, CGM);
 
-    if (done.count(name))
+    if (done.count(mangledName))
       continue;
 
-    LLVM_DEBUG(llvm::dbgs() << "\n-- FUNCTION BEING EMITTED: "
-                            << FD->getNameAsString() << " --\n\n";);
+    LLVM_DEBUG({
+      StringRef prefix = FD->hasAttr<clang::SYCLKernelAttr>()   ? "SYCL KERNEL"
+                         : FD->hasAttr<clang::SYCLDeviceAttr>() ? "SYCL DEVICE"
+                                                                : "";
```

Comment:
I think we should add the "SYCL DEVICE" prefix to functions called from kernels without the `sycl_device` attribute.

Reply:
OK, maybe we can print both the attribute and the context here.